File: apps/user/src/dao/ConnectionGroupDao.ts
'use strict'

import {Db, MongoServerError} from "mongodb";
const { v4: uuidv4 } = require('uuid');

const { MongoClient } = require('mongodb');
const client = new MongoClient(process.env.MONGO_URI);

// models
const ConnectionGroupDocument = require('./document/ConnectionGroupDocument.ts')
import {
    ConnectionGroupExistsError,
} from "../error/Errors";
import {ConnectionGroup} from "../model/connection/ConnectionGroup";
import {SearchResponse} from "../model/response/SearchResponse";
import {SlimUser} from "../model/user/SlimUser";
const UserDocument = require('./document/UserDocument.ts')

class ConnectionGroupDao {
    private db: Db;
    private collectionName: string;

    constructor() {
        this.db = client.db('nevvi')
        this.collectionName = 'connection_groups'
    }

    async createConnectionGroup(userId: string, name: string): Promise<ConnectionGroup> {
        const now = new Date().toISOString()
        const document = new ConnectionGroupDocument({
            id: uuidv4(),
            userId,
            name,
            connections: [],
            createDate: now,
            createBy: 'HARDCODED_FOR_NOW',
            updateDate: now,
            updateBy: 'HARDCODED_FOR_NOW'
        })

        try {
            await this.db.collection(this.collectionName).insertOne(document)
        } catch (e: any) {
            if (e instanceof MongoServerError) {
                throw new ConnectionGroupExistsError(name)
            }
            throw e
        }

        const group = new ConnectionGroup(document)
        group.id = document._id
        return group
    }

    async getConnectionGroup(userId: string, groupId: string): Promise<ConnectionGroup | null> {
        const result = await this.db.collection(this.collectionName)
            .findOne({ userId: userId, _id: groupId })

        if (!result) {
            return null
        }

        const group = new ConnectionGroup(result)
        // @ts-ignore
        group.id = i._id
        return group
    }

    async getConnectionGroups(userId: string): Promise<ConnectionGroup[]> {
        const results = await this.db.collection(this.collectionName)
            .find({ userId: userId })
            .toArray()

        return (results || []).map(i => {
            const group = new ConnectionGroup(i)
            // @ts-ignore
            group.id = i._id
            return group
        })
    }

    async deleteConnectionGroup(userId: string, groupId: string): Promise<boolean> {
        const result = await this.db.collection(this.collectionName)
            .deleteOne({ userId: userId, _id: groupId })

        return result.deletedCount === 1
    }

    async addUserToGroup(userId: string, groupId: string, connectedUserId: string): Promise<boolean> {
        const result = await this.db.collection(this.collectionName)
            .updateOne(
                { userId: userId, _id: groupId },
                { $push: { connections: connectedUserId }}
            )

        return result.modifiedCount === 1
    }

    async removeUserFromGroup(userId: string, groupId: string, connectedUserId: string): Promise<boolean> {
        const result = await this.db.collection(this.collectionName)
            .updateOne(
                { userId: userId, _id: groupId },
                { $pull: { connections: connectedUserId }}
            )

        return result.modifiedCount === 1
    }

    async getConnections(userId: string, groupId: string, name: string | undefined, limit: number, skip: number): Promise<SearchResponse> {
        const pipeline: any = [
            {
                '$match': {
                    'userId': userId,
                    '_id': groupId,
                }
            },
            {
                '$unwind': {
                    'path': '$connections',
                    'preserveNullAndEmptyArrays': false
                }
            },
            {
                '$lookup': {
                    'from': 'users',
                    'localField': 'connections',
                    'foreignField': '_id',
                    'as': 'connectedUser'
                }
            }
        ]

        if (name) {
            const search = name.split(' ').filter(n => n).join('_').toLowerCase()
            pipeline.push({
                '$match': {
                    'connectedUser.nameLower':  {$regex : search}
                }
            })
        }

        pipeline.push({
            '$sort': {
                "connectedUser.lastName": 1,
                "connectedUser.firstName": 1
            }
        })

        pipeline.push({
            '$facet': {
                'connections': [
                    {
                        '$skip': skip
                    },
                    {
                        '$limit': limit
                    }
                ],
                'connectionCount': [
                    {
                        '$count': 'connectionCount'
                    }
                ]
            }
        })

        const result = await this.db.collection(this.collectionName)
            .aggregate(pipeline)
            .next()

        // Return the user that we mapped over so that we don't need to grab all that info again 1 by 1...
        // Also return the total count before pagination was applied
        const userResults: any[] = (result?.connections || [])
        const users = userResults
            .filter(i => i["connectedUser"] && i["connectedUser"].length === 1)
            .map(i => {
                const user = new SlimUser(new UserDocument(i["connectedUser"][0]))
                user.id = i["connectedUser"][0]._id
                user.connected = true // if this user is in a group they darn will be connected
                return user
            })

        // connectionCount comes back nested
        let userCount = 0
        if (result && result.connectionCount && result.connectionCount.length === 1) {
            userCount = result.connectionCount[0].connectionCount || 0
        }

        return new SearchResponse(users, userCount)
    }
}

export {ConnectionGroupDao}

File: apps/user/src/dao/connection_group_dao.py
import os
import uuid
from datetime import datetime, timezone
from typing import List, Optional, Dict, Any

import pymongo
from pymongo.errors import DuplicateKeyError
from pymongo.synchronous.collection import Collection

from src.dao.user_dao import SearchedUser
from src.model.document import ConnectionGroupDocument, ConnectionGroupSearch
from src.model.errors import ConnectionGroupExistsError


class ConnectionGroupDao:
    def __init__(self):
        client = pymongo.MongoClient(os.environ.get("MONGO_URI"))
        self.collection: Collection[ConnectionGroupDocument] = client.get_database("nevvi").get_collection(
            "connection_groups")

    def create_group(self, user_id: str, name: str) -> ConnectionGroupDocument:
        now = datetime.now(timezone.utc).isoformat()
        document = ConnectionGroupDocument(
            _id=str(uuid.uuid4()),
            name=name,
            userId=user_id,
            connections=[],
            createDate=now,
            updateDate=now,
        )

        try:
            self.collection.insert_one(document)
        except DuplicateKeyError:
            raise ConnectionGroupExistsError(name)

        return document

    def get_group(self, user_id: str, group_id: str) -> Optional[ConnectionGroupDocument]:
        return self.collection.find_one(filter={
            "userId": user_id,
            "_id": group_id
        })

    def get_groups(self, user_id: str) -> List[ConnectionGroupDocument]:
        return list(self.collection.find(filter={
            "userId": user_id,
        }))

    def delete_group(self, user_id: str, group_id: str) -> bool:
        res = self.collection.delete_one({
            "userId": user_id,
            "_id": group_id
        })

        return res.deleted_count == 1

    def add_user(self, user_id: str, group_id: str, connected_user_id: str) -> bool:
        res = self.collection.update_one(
            filter={"userId": user_id, "_id": group_id},
            update={"$push": {"connections": connected_user_id}}
        )

        return res.modified_count == 1

    def remove_user(self, user_id: str, group_id: str, connected_user_id: str) -> bool:
        res = self.collection.update_one(
            filter={"userId": user_id, "_id": group_id},
            update={"$pull": {"connections": connected_user_id}}
        )

        return res.modified_count == 1

    def get_connections(self, user_id: str, group_id: str, limit: int, skip: int,
                        name: Optional[str] = None) -> ConnectionGroupSearch:
        pipeline = [
            {
                '$match': {
                    'userId': user_id,
                    '_id': group_id,
                }
            },
            {
                '$unwind': {
                    'path': '$connections',
                    'preserveNullAndEmptyArrays': False
                }
            },
            {
                '$lookup': {
                    'from': 'users',
                    'localField': 'connections',
                    'foreignField': '_id',
                    'as': 'connectedUser'
                }
            }
        ]

        if name is not None:
            search = '_'.join(name.split(' ')).lower()
            pipeline.append({
                '$match': {
                    'connectedUser.nameLower': {"$regex": search}
                }
            })

        pipeline.append({
            '$sort': {
                "connectedUser.lastName": 1,
                "connectedUser.firstName": 1
            }
        })

        pipeline.append({
            '$facet': {
                'connections': [
                    {
                        '$skip': skip
                    },
                    {
                        '$limit': limit
                    }
                ],
                'connectionCount': [
                    {
                        '$count': 'connectionCount'
                    }
                ]
            }
        })

        res: Dict[str, Any] = self.collection.aggregate(pipeline).next()

        def transform(result: Dict[str, Any]) -> SearchedUser:
            return SearchedUser(
                id=result.get("_id"),
                firstName=result.get("firstName"),
                lastName=result.get("lastName"),
                bio=result.get("bio"),
                profileImage=result.get("profileImage"),
                phoneNumber=result.get("phoneNumber"),
                connected=True,
                requested=True
            )

        connections = [transform(conn.get("connectedUser")[0]) for conn in res.get("connections", []) if
                       len(conn.get("connectedUser", [])) == 1]

        user_count = 0
        if res and len(res.get("connectionCount", [])) == 1:
            user_count = res.get("connectionCount")[0].get("connectionCount", 0)

        return ConnectionGroupSearch(connections, user_count)


if __name__ == '__main__':
    os.environ["MONGO_URI"] = "REPLACEME"
    dao = ConnectionGroupDao()

    # dao.create_group(user_id="74384478-f0c1-7059-097f-2555f8ff7a2c", name="Test2")
    # print(dao.get_group(user_id="74384478-f0c1-7059-097f-2555f8ff7a2c", group_id="362e154c-36d5-477f-b5a4-da7478caf3d9"))
    # print(dao.get_groups(user_id="74384478-f0c1-7059-097f-2555f8ff7a2c"))
    # dao.add_user(user_id="74384478-f0c1-7059-097f-2555f8ff7a2c", group_id="362e154c-36d5-477f-b5a4-da7478caf3d9", connected_user_id="d4d84418-40b1-7042-8efd-7a121430a882")
    # print(dao.get_connections(user_id="74384478-f0c1-7059-097f-2555f8ff7a2c", group_id="362e154c-36d5-477f-b5a4-da7478caf3d9", limit=10, skip=0))
